### PR TITLE
ETW input - session name prefix and old session cleanup

### PR DESCRIPTION
**Pull request by kwaclaw**
This pull request is in response to issue https://github.com/Azure/diagnostics-eventflow/issues/232 .

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -441,6 +444,9 @@ To capture data from EventSources running in the same process as EventFlow, the
 | Field | Values/Types | Required | Description |
 | :---- | :-------------- | :------: | :---------- |
 | `type` | "ETW" | Yes | Specifies the input type. For this input, it must be "ETW". |
+| `sessionNamePrefix` | string | No | The ETW trace session will be created with this name prefix, which helps in differentiating  ETW input instances owned by multiple applications. If not set, a default prefix will be used. |
+| `cleanupOldSessions` | boolean | No | If set, existing ETW trace sessions matching the `sessionNamePrefix` will be closed. This helps to collect leftover session instances, as there is a limit on their number. |
+| `restartExisting` | boolean | No | If turned on, then an existing trace session matching the `sessionNamePrefix` will be re-used. If `cleanupOldSessions` is also turned on, then it will leave one session open for re-use. |
```
Comment (karolz-ms) [inline README.md]: Please rename the setting to `reuseExistingSession`

File: src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/EtwInput.cs [hunk 2]
```diff
@@ -24,12 +24,17 @@ public class EtwInput : IObservable<EventData>, IDisposable, IRequireActivation
         private EventFlowSubject<EventData> subject;
         private ITraceEventSession session;
         private bool isDisposed;
+        private readonly string defaultSessionnamePrefix = $"EventFlow-{nameof(EtwInput)}";
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/EtwInput.cs]: Make it `private static readonly string DefaultSessionNamePrefix` and move up next to `ProviderSectionName`

File: src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/StandardTraceEventSession.cs [hunk 3]
```diff
@@ -20,15 +20,46 @@ internal class StandardTraceEventSession : ITraceEventSession
...
+            }
+            if (sessionName == null)
+                sessionName = $"{sessionNamePrefix }-{Guid.NewGuid().ToString()}";
+
+            // even if the session already exists, we must restart it as we cannot enable providers on an attached session
+            this.inner = new TraceEventSession(sessionName, TraceEventSessionOptions.Create);
             this.isProcessing = false;
         }
 
+        string CleanupMatchingSessions(string sessionNamePrefix, bool keepOne)
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/StandardTraceEventSession.cs]: private

Please move private method to after all public methods

File: src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/StandardTraceEventSession.cs [hunk 4]
```diff
@@ -20,15 +20,46 @@ internal class StandardTraceEventSession : ITraceEventSession
...
+        {
+            string result = null;
+
+            foreach (var sesName in TraceEventSession.GetActiveSessionNames())
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/StandardTraceEventSession.cs]: I do not see the `keepOne` parameter being used--looks like a session is always reused?

Please rename the parameter to `reuseExistingSession` to keep it consistent with configuration

File: src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/StandardTraceEventSession.cs [hunk 5]
```diff
@@ -20,15 +20,46 @@ internal class StandardTraceEventSession : ITraceEventSession
...
             this.healthReporter = healthReporter;
-            this.inner = new TraceEventSession($"EventFlow-{nameof(EtwInput)}-{Guid.NewGuid().ToString()}", TraceEventSessionOptions.Create);
+
+            string sessionName = null;
+            if (cleanupOldSessions)
+            {
+                sessionName = CleanupMatchingSessions(sessionNamePrefix, restartExisting);
+            }
+            if (sessionName == null)
+                sessionName = $"{sessionNamePrefix }-{Guid.NewGuid().ToString()}";
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/StandardTraceEventSession.cs]: (nitpick)

In EventFlow codebase we use curly braces for `if` statement body even if there is only one statement in the body, see lines 31-34 for example. For the sake of consistency please follow this convention--thanks!

(applies to `CleanupMatchingSessions` method too)

**Comment by kwaclaw** [comment]
I think I have addressed all your comments.

**Comment by karolz-ms** [comment]
Cool, thank you Karl, much appreciated. I will publish the Nuget package tomorrow

**Comment by kwaclaw** [comment]
Thanks for the quick turnaround!

**Comment by karolz-ms** [comment]
The 1.3.2 package is now available from Nuget. Cheers!

**Pull request merged**